meta-balena-common/recipes-connectivity/networkmanager/resin-files/98dhcp_ntp

```diff
@@ -0,0 +1,49 @@
+#!/bin/sh
+
+#
+# Update NTP sources that are passed via DHCP.
+#
+
+INTERFACE=$1
+ACTION=$2
+
+SERVER_DIR=/run/chrony/
+SERVER_FILE=${SERVER_DIR}/added_dhcp_${INTERFACE}.sources
+
+if [ ! -d "$SERVER_DIR" ]; then
+	mkdir -p $SERVER_DIR
+	chmod 750 $SERVER_DIR
+fi
+
+case "$ACTION" in
+	up|dhcp4-change)
+	if [[ -n $DHCP4_NTP_SERVERS ]]; then
+		if [ -f "$SERVER_FILE" ]; then
+			echo "Updating DHCP NTP sources (${INTERFACE} ${ACTION})."
+			rm -f $SERVER_FILE
+		else
+			echo "Adding DHCP NTP sources (${INTERFACE} ${ACTION})."
+		fi
+
+		for server in ${DHCP4_NTP_SERVERS}; do
+			echo "pool $server iburst minpoll 14 maxpoll 14 maxsources 1" >> $SERVER_FILE
+		done
+		/usr/libexec/chrony-helper update-sources || true
```

Comment:
If chrony is not running when we run this it will silently fail as the chrony-helper script returns success in that case. Is there a way we could at least print a warning message and at best make sure chrony is run and retry?

Reply:
The new update method means that we don't actually care whether chrony is running at the point we receive the NTP sources via DHCP. This allows us to track changes in NTP sources regardless of whether chrony has been stopped (by the user etc.) or whether it is started much later in the boot process.

The DHCP NTP sources are written to a '.sources' file in /run/chrony/ when they are received. If chrony is running then chrony-helper triggers a 'reload sources' to force chrony to update it's source list from the '.sources' files. If chrony is not running then the '.sources' files will automatically be read when chrony is (re)started.

This is the same technique that is used for chrony by Ubuntu and also the same as the example 'chrony.nm-dispatcher.dhcp' script in the chrony repository.

Reply:
thanks for the explanation @markcorbinuk